yjit/src/codegen.rs

```diff
@@ -2156,26 +2157,19 @@ fn gen_defined(
 
     // Save the PC and SP because the callee may allocate
     // Note that this modifies REG_SP, which is why we do it first
-    jit_prepare_routine_call(jit, ctx, cb, REG0);
+    jit_prepare_routine_call(jit, ctx, asm);
 
     // Get the operands from the stack
     let v_opnd = ctx.stack_pop(1);
 
     // Call vm_defined(ec, reg_cfp, op_type, obj, v)
-    mov(cb, C_ARG_REGS[0], REG_EC);
-    mov(cb, C_ARG_REGS[1], REG_CFP);
-    mov(cb, C_ARG_REGS[2], uimm_opnd(op_type.into()));
-    jit_mov_gc_ptr(jit, cb, C_ARG_REGS[3], obj);
-    mov(cb, C_ARG_REGS[4], v_opnd);
-    call_ptr(cb, REG0, rb_vm_defined as *const u8);
+    let def_result = asm.ccall(rb_vm_defined as *const u8, vec![EC, CFP, op_type.into(), obj.into(), v_opnd]);
 
     // if (vm_defined(ec, GET_CFP(), op_type, obj, v)) {
     //  val = pushval;
     // }
-    jit_mov_gc_ptr(jit, cb, REG1, pushval);
-    cmp(cb, AL, imm_opnd(0));
-    mov(cb, RAX, uimm_opnd(Qnil.into()));
-    cmovnz(cb, RAX, REG1);
+    asm.cmp(def_result, Opnd::Imm(0));
+    let out_value = asm.csel_nz(pushval.into(), Qnil.into());
```

Comment:
I think this could be the cause of the failure.

The previous code was comparing AL (lowest 8-bit) against 0. The new comparison is comparing the whole 64-bit. The problem is that C booleans only use the lowest 8 bits, so we can't be sure what the higher bits are. Possibly we'd want to implement an `Opnd::sub_reg(num_bits)` method for the `Opnd` struct (this method already exists for x86 and arm).

Before proceeding with this, I'd print out the values of `def_result`, check if the higher bits are ever nonzero. There's a `print_int(asm, opnd)` function that you can use to print out `def_result`.